File: Haskell/Problem2.hs
module ProjectEuler.Problem2 where

sumEvenFibonacciNumbers :: Integral a => a -> a
sumEvenFibonacciNumbers m = sum $ takeWhile (<= m) evenFibs
  where
    evenFibs = 2 : 8 : zipWith (\a b -> a + 4 * b) evenFibs (tail evenFibs)

main :: IO ()
main = print $ sumEvenFibonacciNumbers 4000000

File: Scala/Problem2.scala
object Problem2 {
  def sumEvenFibonacciNumbers(m: BigInt): BigInt = {
    lazy val evenFibs: Stream[BigInt] =
      2 #:: 8 #:: evenFibs.zip(evenFibs.tail).map(t => t._1 + 4 * t._2)
    evenFibs.takeWhile(_ <= m).sum
  }

  def main(args: Array[String]): Unit =
    println(sumEvenFibonacciNumbers(4000000))
}
